core-bundle/tests/Migration/Version40900/CeAccessMigrationTest.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Tests\Migration\Version40900;
+
+use Contao\ContentText;
+use Contao\CoreBundle\Migration\Version40900\CeAccessMigration;
+use Contao\CoreBundle\Tests\TestCase;
+use Contao\FormTextField;
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Schema\MySqlSchemaManager;
+use Doctrine\DBAL\Statement;
+
+class CeAccessMigrationTest extends TestCase
+{
+    public static function setUpBeforeClass(): void
+    {
+        parent::setUpBeforeClass();
+
+        $GLOBALS['TL_CTE'] = [
+            'texts' => [
+                'text' => ContentText::class,
+            ],
+        ];
+
+        $GLOBALS['TL_FFL'] = [
+            'text' => FormTextField::class,
+        ];
+    }
+
+    public static function tearDownAfterClass(): void
+    {
+        parent::tearDownAfterClass();
+
+        unset($GLOBALS['TL_CTE'], $GLOBALS['TL_FFL']);
+    }
+
+    public function testActivatesTheFieldsInAllUserGroups(): void
+    {
+        $schemaManager = $this->createMock(MySqlSchemaManager::class);
+        $schemaManager
+            ->expects($this->once())
+            ->method('tablesExist')
+            ->with(['tl_user_group'])
+            ->willReturn(true)
+        ;
+
+        $schemaManager
+            ->expects($this->once())
+            ->method('listTableColumns')
+            ->willReturn([])
+        ;
+
+        $stmt = $this->createMock(Statement::class);
+        $stmt
+            ->expects($this->once())
+            ->method('execute')
+            ->with([':elements' => 'a:1:{i:0;s:4:"text";}', ':fields' => 'a:1:{i:0;s:4:"text";}'])
+        ;
+
+        $connection = $this->createMock(Connection::class);
+        $connection
+            ->expects($this->once())
+            ->method('getSchemaManager')
+            ->willReturn($schemaManager)
+        ;
+
+        $connection
+            ->expects($this->once())
+            ->method('query')
+        ;
+
+        $connection
+            ->expects($this->once())
+            ->method('prepare')
+            ->willReturn($stmt)
+        ;
+
+        $migration = new CeAccessMigration($connection);
+
+        $this->assertTrue($migration->shouldRun());
+        $this->assertTrue($migration->run()->isSuccessful());
+    }
+
+    public function testDoesNothingIfTheElementsColumnExists(): void
```

Comment:
Shouldn’t this be name somthing like this?

```suggestion
    public function testDoesNothingIfTheUserGroupTableDoesNotExist(): void
```

Reply:
Fixed in ddf16273efefc782475adf69b024ffb63ae5d6f7.

I am not sure how much sense these tests make, though. (I am talking about the whole class, not just the two methods added/fixed in the most recent commit.)

Reply:
I’m not sure either. Should we make a functional test instead that migrates a real database?

Reply:
I am not sure. The unit test already ensures that the correct query parameters are used, so a function tests would basically only test that the application can write to the database. 🤷‍♂